pkg/client/oci/oci_test.go

```diff
@@ -0,0 +1,209 @@
+package oci_test
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path"
+	"reflect"
+	"sort"
+	"testing"
+	"time"
+
+	"github.com/bitnami-labs/charts-syncer/api"
+	"github.com/bitnami-labs/charts-syncer/internal/cache"
+	"github.com/bitnami-labs/charts-syncer/internal/utils"
+	"github.com/bitnami-labs/charts-syncer/pkg/client/oci"
+	"github.com/bitnami-labs/charts-syncer/pkg/client/types"
+	"github.com/docker/distribution/configuration"
+	"github.com/docker/distribution/registry"
+	_ "github.com/docker/distribution/registry/storage/driver/inmemory"
+	"github.com/phayes/freeport"
+	"helm.sh/helm/v3/pkg/chart"
+)
+
+var (
+	ociRepo = &api.Repo{
+		Kind: api.Kind_OCI,
+		Auth: &api.Auth{
+			Username: "user",
+			Password: "password",
+		},
+	}
+)
+
+func prepareTest(t *testing.T) *oci.Repo {
+	t.Helper()
+
+	// Define cache dir
+	cacheDir, err := ioutil.TempDir("", "client")
+	if err != nil {
+		t.Fatal(err)
+	}
+	cache, err := cache.New(cacheDir, ociRepo.GetUrl())
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(func() { os.RemoveAll(cacheDir) })
+
+	// Create oci client
+	client, err := oci.New(ociRepo, cache)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return client
+}
+
+// Creates an HTTP server that knows how to reply to all OCI related request except PUSH one.
+func prepareHttpServer(t *testing.T) *oci.Repo {
+	t.Helper()
+
+	// Create HTTP server
+	tester := oci.NewTester(t, ociRepo)
+	ociRepo.Url = tester.GetURL() + "/someproject/charts"
+
+	return prepareTest(t)
+}
+
+// Starts an OCI compliant server (docker-registry) so our push command based on oras cli works out-of-the-box.
+// This way we don't have to mimic all the low-level HTTP requests made by oras.
+func prepareOciServer(t *testing.T) *oci.Repo {
+	t.Helper()
+
+	// Create OCI server as docker registry
+	config := &configuration.Configuration{}
+	port, err := freeport.GetFreePort()
```

Comment:
You don't need to use an external package for this. See the example below and feel free to add it to the `utils` package.

```
func getListenAddress(t *testing.T) string {
	t.Helper()

	lst, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatal(err)
	}
	defer lst.Close()

	return lst.Addr().String()
}
```